File: src/GPASim_06_ABC.r
##############################
### ANALYSIS PER LANDSCAPE ### MODULE 3 of 3:
############################## ABC posterior distribution estimation via aceptance-rejection algorithm

####################
### user inputs ###
args = commandArgs(trailingOnly=TRUE)
DIR = args[1]   ### GPASIM output directory
PLOT = args[2]  ### save plots?
# ### TEST:
# DIR = getwd()
# PLOT = TRUE

#############################
### set working directory ###
setwd(DIR)

###########################################################
### extract landscape parameters from th directory name ###
REP = as.numeric(sub("rep", "", unlist(strsplit(basename(DIR), "_"))[2]))   ### replication number
QTL = as.numeric(sub("qtl", "", unlist(strsplit(basename(DIR), "_"))[3]))   ### number of QTL
MGR = as.numeric(sub("mr", "", unlist(strsplit(basename(DIR), "_"))[4]))    ### migration rate
FGS = as.numeric(sub("fgs", "", unlist(strsplit(basename(DIR), "_"))[5]))   ### foreground selection intensity
BGS = as.numeric(sub("bgs", "", unlist(strsplit(basename(DIR), "_"))[6]))   ### backegroun selection intensity
GRAD = as.numeric(sub("grad", "", unlist(strsplit(basename(DIR), "_"))[7])) ### resistance alleles landscape gradient

######################
### load libraries ###
library(abc)
library(RColorBrewer)

##########################################
### define observed summary statistics ###
### i.e. metrics we want to optimize: correlation, log10(RMSD +1), true positive rate and false positive rate
TARGET_VECTOR = c(CORRELATION=1.0,
                  LOG10_RMSD=0.0,
                  TRUE_POSITIVE_RATE=1.0,
                  FALSE_POSITIVE_RATE=0.0)

##########################################
### load the simulated data from Julia ###
### header: REP, ACROSS_POOL, ACROSS_INDI, WITHIN_POOL, WITHIN_INDI, CORRELATION, LOG10_RMSD, TRUE_POSITIVE_RATE, FALSE_POSITIVE_RATE
# DATA = read.csv("ABC_SAMPLING_OUTPUT_1000REPS.csv")
DATA = read.csv("ABC_SAMPLING_OUTPUT_100REPS.csv")
# DATA = read.csv("ABC_SAMPLING_OUTPUT_10REPS.csv")
### heuristically determine an appropriate tolerance value via odd-one-out cross-validation
start = Sys.time()
cv = cv4abc(param=DATA[,2:5], sumstat=DATA[,6:9], nval=5, tols=c(0.001, 0.01, 0.1), statistic="mean", method="rejection") ### 3.8 hours!
end = Sys.time()
print(end - start)
tol = as.numeric(names(which.min(rowMeans(scale(summary(cv))))))
### perform ABC rejection algorithm and estimate the posterior distribution of the parameters
out = abc(target=TARGET_VECTOR, param=DATA[,2:5], sumstat=DATA[,6:9], tol=tol, method="rejection")
DATA_SELECTED = DATA[out$region,]

#####################################################################################################
### plot summary statistics distribution and the posterior distribution of the summary statistics ###
if (PLOT==TRUE){svg(paste0(basename(DIR), "-ABC_resource_optim.svg"), width=6, height=10)}
par(mfrow=c(4,2))
LABELS = c("Across Populations & Pool-Level",
           "Across Populations & Individual-Level",
           "Within Populations & Pool-Level",
           "Within Populations & Individual-Level",
           "Correlation",
           "log10(RMSD + 1)",
           "True Positive Rate",
           "False Positive Rate")
ncolors = length(LABELS)
colors_fill = rep(brewer.pal(min(9, ncolors), "Pastel1"), times=ceiling(ncolors/9))
colors_bord = rep(brewer.pal(min(9, ncolors), "Set1"), times=ceiling(ncolors/9))
for (i in 1:length(LABELS)){
  hist(DATA_SELECTED[, (i+1)], main=LABELS[i], xlab=colnames(DATA_SELECTED)[i+1], col=colors_fill[i], border=colors_bord[i])
  legend("topright", legend=round(mean(DATA_SELECTED[, (i+1)]),2))
}
if (PLOT==TRUE){dev.off()}

############################
### plot NULL parameters ### equal partitioning
nREP = max(DATA$REP)
nLIB = 10 ### same as the number of librries used in "assessing_CROSS_VALIDATION_OUTPUT_MERGED_2.jl"
NULL_SUBSET = droplevels(DATA[(DATA$ACROSS_POOL==0.25) & (DATA$ACROSS_INDI==0.25) & (DATA$WITHIN_POOL==0.25) & (DATA$WITHIN_INDI==0.25), ])
NULL_CORRELATION = c()
NULL_LOG10_RMSD = c()
NULL_TRUE_POSITIVE_RATE = c()
NULL_FALSE_POSITIVE_RATE = c()
for (i in 1:nREP){
  idx = sample(1:nrow(NULL_SUBSET), nLIB, replace=FALSE)
  NULL_CORRELATION = c(NULL_CORRELATION, mean(NULL_SUBSET$CORRELATION[idx], na.rm=TRUE))
  NULL_LOG10_RMSD = c(NULL_LOG10_RMSD, mean(NULL_SUBSET$LOG10_RMSD[idx], na.rm=TRUE))
  NULL_TRUE_POSITIVE_RATE = c(NULL_TRUE_POSITIVE_RATE, mean(NULL_SUBSET$TRUE_POSITIVE_RATE[idx], na.rm=TRUE))
  NULL_FALSE_POSITIVE_RATE = c(NULL_FALSE_POSITIVE_RATE, mean(NULL_SUBSET$FALSE_POSITIVE_RATE[idx], na.rm=TRUE))
}
NULL_PARTITION = rep(0.25, times=nREP)
DATA_NULL = data.frame(REP=1:nREP,
                       ACROSS_POOL=NULL_PARTITION,
                       ACROSS_INDI=NULL_PARTITION,
                       WITHIN_POOL=NULL_PARTITION,
                       WITHIN_INDI=NULL_PARTITION,
                       CORRELATION = NULL_CORRELATION,
                       LOG10_RMSD = NULL_LOG10_RMSD,
                       TRUE_POSITIVE_RATE = NULL_TRUE_POSITIVE_RATE,
                       FALSE_POSITIVE_RATE = NULL_FALSE_POSITIVE_RATE)
if (PLOT==TRUE){svg(paste0(basename(DIR), "-ABC_NULL.svg"), width=6, height=10)}
par(mfrow=c(4,2))
LABELS = c("Across Populations & Pool-Level",
           "Across Populations & Individual-Level",
           "Within Populations & Pool-Level",
           "Within Populations & Individual-Level",
           "Correlation",
           "log10(RMSD + 1)",
           "True Positive Rate",
           "False Positive Rate")
ncolors = length(LABELS)
colors_fill = rep(brewer.pal(min(9, ncolors), "Pastel1"), times=ceiling(ncolors/9))
colors_bord = rep(brewer.pal(min(9, ncolors), "Set1"), times=ceiling(ncolors/9))
for (i in 1:length(LABELS)){
  hist(DATA_NULL[, (i+1)], main=LABELS[i], xlab=colnames(DATA_NULL)[i+1], col=colors_fill[i], border=colors_bord[i])
  legend("topright", legend=round(mean(DATA_NULL[, (i+1)]),2))
}
if (PLOT==TRUE){dev.off()}

###########################################################################################################################
### Separate GP and GWAS ABC (using the tolerance value derived from the ABC using both GWAS and GP summary statistics) ###
out_GP   = abc(target=c(1.0, 0.0), param=DATA[,2:5], sumstat=DATA[,6:7], tol=tol, method="rejection")
out_GWAS = abc(target=c(1.0, 0.0), param=DATA[,2:5], sumstat=DATA[,8:9], tol=tol, method="rejection")
summary(out_GP)
summary(out_GWAS)
DATA_SELECTED_GP = DATA[out_GP$region,]
DATA_SELECTED_GWAS = DATA[out_GWAS$region,]

### plot summary statistics distribution and the posterior distribution of the summary statistics ###
if (PLOT==TRUE){svg(paste0(basename(DIR), "-ABC_PER_GP_GWAS.svg"), width=8, height=10)}
par(mfrow=c(6,2))
LABELS_GP = c("Across Populations & Pool-Level",
           "Across Populations & Individual-Level",
           "Within Populations & Pool-Level",
           "Within Populations & Individual-Level",
           "Correlation",
           "log10(RMSD + 1)")
LABELS_GWAS = c("Across Populations & Pool-Level",
          "Across Populations & Individual-Level",
          "Within Populations & Pool-Level",
          "Within Populations & Individual-Level",
          "True Positive Rate",
          "False Positive Rate")
ncolors = length(LABELS_GP) + length(LABELS_GWAS)
colors_fill = rep(brewer.pal(min(9, ncolors), "Pastel1"), times=ceiling(ncolors/9))
colors_bord = rep(brewer.pal(min(9, ncolors), "Set1"), times=ceiling(ncolors/9))
for (i in 1:length(LABELS_GP)){
  hist(DATA_SELECTED_GP[, (i+1)], main=LABELS_GP[i], xlab=colnames(DATA_SELECTED_GP)[i+1], col=colors_fill[i], border=colors_bord[i])
  legend("topright", legend=round(mean(DATA_SELECTED_GP[, (i+1)]),2))
}
for (i in 1:length(LABELS_GWAS)){
  hist(DATA_SELECTED_GWAS[, (i+1)], main=LABELS_GWAS[i], xlab=colnames(DATA_SELECTED_GWAS)[i+1], col=colors_fill[i], border=colors_bord[i])
  legend("topright", legend=round(mean(DATA_SELECTED_GWAS[, (i+1)]),2))
}
if (PLOT==TRUE){dev.off()}

######################################################################################################################################################################################
###################
### MAIN OUTPUT ###
###################
### ABC partitioning and metrics summary statistics for each ABC_CLASS testing the:
### (1) NULL hyposthesis,
### (2) optmizing for GP (corr and RMSD) only,
### (3) optimizing for GWAS (tpr and fpr) only, and
### (4) optimizing for both GP and GWAS simultaneously.
OUT = data.frame(REP = rep(REP, each=4),
                 QTL = rep(QTL, each=4),
                 MGR = rep(MGR, each=4),
                 FGS = rep(FGS, each=4),
                 BGS = rep(BGS, each=4),
                 GRAD = rep(GRAD, each=4),
                 ABC_CLASS = c("NULL", "GP", "GWAS", "GPAS"),
                 ACROSS_POOL = c(mean(DATA_NULL$ACROSS_POOL, na.rm=TRUE), mean(DATA_SELECTED_GP$ACROSS_POOL, na.rm=TRUE), mean(DATA_SELECTED_GWAS$ACROSS_POOL, na.rm=TRUE), mean(DATA_SELECTED$ACROSS_POOL, na.rm=TRUE)),
                 ACROSS_INDI = c(mean(DATA_NULL$ACROSS_INDI, na.rm=TRUE), mean(DATA_SELECTED_GP$ACROSS_INDI, na.rm=TRUE), mean(DATA_SELECTED_GWAS$ACROSS_INDI, na.rm=TRUE), mean(DATA_SELECTED$ACROSS_INDI, na.rm=TRUE)),
                 WITHIN_POOL = c(mean(DATA_NULL$WITHIN_POOL, na.rm=TRUE), mean(DATA_SELECTED_GP$WITHIN_POOL, na.rm=TRUE), mean(DATA_SELECTED_GWAS$WITHIN_POOL, na.rm=TRUE), mean(DATA_SELECTED$WITHIN_POOL, na.rm=TRUE)),
                 WITHIN_INDI = c(mean(DATA_NULL$WITHIN_INDI, na.rm=TRUE), mean(DATA_SELECTED_GP$WITHIN_INDI, na.rm=TRUE), mean(DATA_SELECTED_GWAS$WITHIN_INDI, na.rm=TRUE), mean(DATA_SELECTED$WITHIN_INDI, na.rm=TRUE)),
                 CORRELATION = c(mean(DATA_NULL$CORRELATION, na.rm=TRUE), mean(DATA_SELECTED_GP$CORRELATION, na.rm=TRUE), mean(DATA_SELECTED_GWAS$CORRELATION, na.rm=TRUE), mean(DATA_SELECTED$CORRELATION, na.rm=TRUE)),
                 LOG10_RMSD = c(mean(DATA_NULL$LOG10_RMSD, na.rm=TRUE), mean(DATA_SELECTED_GP$LOG10_RMSD, na.rm=TRUE), mean(DATA_SELECTED_GWAS$LOG10_RMSD, na.rm=TRUE), mean(DATA_SELECTED$LOG10_RMSD, na.rm=TRUE)),
                 TRUE_POSITIVE_RATE = c(mean(DATA_NULL$TRUE_POSITIVE_RATE, na.rm=TRUE), mean(DATA_SELECTED_GP$TRUE_POSITIVE_RATE, na.rm=TRUE), mean(DATA_SELECTED_GWAS$TRUE_POSITIVE_RATE, na.rm=TRUE), mean(DATA_SELECTED$TRUE_POSITIVE_RATE, na.rm=TRUE)),
                 FALSE_POSITIVE_RATE = c(mean(DATA_NULL$FALSE_POSITIVE_RATE, na.rm=TRUE), mean(DATA_SELECTED_GP$FALSE_POSITIVE_RATE, na.rm=TRUE), mean(DATA_SELECTED_GWAS$FALSE_POSITIVE_RATE, na.rm=TRUE), mean(DATA_SELECTED$FALSE_POSITIVE_RATE, na.rm=TRUE)))
write.table(OUT, file=paste0("ABC_RESOURCE_OPTIM-", REP, "rep-", QTL, "qtl-", MGR, "mgr-", FGS, "fgs-", BGS, "bgs-", GRAD, "grad.csv"), sep=",", quote=FALSE, row.names=FALSE)
######################################################################################################################################################################################

# #####################
# ### MISCELLANEOUS ###
# #####################
# ####################################
# ### plot slected parameter space ###
# par(mfrow=c(2,3))
# for (x in 1:3){
#   for (y in (x+1):4){
#     plot(x=DATA_SELECTED[, x+1], y=DATA_SELECTED[, y+1], xlab=LABELS[x], ylab=LABELS[y], main="", type="p", pch=20, col=c(colors_fill[x], colors_fill[y]))
#   }
# }
#
# library(rgl)
# plot3d(
#   x=DATA_SELECTED$WITHIN_POOL,
#   y=DATA_SELECTED$WITHIN_INDI,
#   z=DATA_SELECTED$TRUE_POSITIVE_RATE,
#   col=rgb(0.1, 0.9, 0.5, alpha=0.5)
# )
# ##################################
# ### lanscape-based model plots ###

File: src/GPASim_06_ABC.jl
##############################
### ANALYSIS PER LANDSCAPE ### MODULE 2 of 3:
############################## ABC simulations via parallel processing

### simulate summary stats across this parameter space for NREP replications
### In Julia
# julia -p12
### inputs
DIR = ARGS[1]   ### GPASIM output directory
nQTL = parse(Int64, ARGS[2]) ### number of QTL
nREP = parse(Int64, ARGS[3]) ### number of replications
# ### TEST:
# DIR = "/data/Lolium/Quantitative_Genetics/LOLSIM_2019_TEST/LOLSIM_1rep_5qtl_0.001mr_0.25fgs_0.00bgs_0grad"
# nQTL = 5
# nREP = 10
### load libraries and modules
using Distributed
Distributed.addprocs(length(Sys.cpu_info()))
@everywhere using Distributed
@everywhere using DataFrames
@everywhere using CSV
@everywhere using DelimitedFiles
@everywhere using Statistics
@everywhere using ProgressMeter
@everywhere using SharedArrays
using ProgressMeter
### set working directory
cd(DIR)
### load the streamlined GPAS cross-validation output
@everywhere ACROSS_POOL = CSV.read("ACROSS_POOL.csv")
@everywhere ACROSS_INDI = CSV.read("ACROSS_INDI.csv")
@everywhere WITHIN_POOL = CSV.read("WITHIN_POOL.csv")
@everywhere WITHIN_INDI = CSV.read("WITHIN_INDI.csv")
### a very inelegant fix to putting nQTL on all threads
@everywhere nQTL = convert(Int64, DelimitedFiles.readdlm("nQTL.temp", ',')[1,1])
### generate the parameter space
@everywhere across_pool = []
@everywhere across_indi = []
@everywhere within_pool = []
@everywhere within_indi = []
epsilon = 0.01
pb = ProgressMeter.Progress(length(collect(0.0:epsilon:1.0))^4, dt=1, desc="Generating parameter space: ",  barglyphs=BarGlyphs("[=> ]"), barlen=50, color=:yellow); counter=[1]
for i in collect(0.0:epsilon:1.0)
    for j in collect(0.0:epsilon:1.0)
        for k in collect(0.0:epsilon:1.0)
            for l in collect(0.0:epsilon:1.0)
                if (i+j+k+l == 1.00)
                    push!(across_pool, i)
                    push!(across_indi, j)
                    push!(within_pool, k)
                    push!(within_indi, l)
                end
                ProgressMeter.update!(pb, counter[1]); counter[1] = counter[1] + 1
            end
        end
    end
end
@everywhere PAR = DataFrames.DataFrame(ACROSS_POOL=across_pool, ACROSS_INDI=across_indi, WITHIN_POOL=within_pool, WITHIN_INDI=within_indi)
### define the sampling_summstats_function
@everywhere function SAMPLE_ABC(;par::Array{Float64,1}, nLib::Int64=10, nQTL::Int64=nQTL, ACROSS_POOL::DataFrames.DataFrame=ACROSS_POOL, ACROSS_INDI::DataFrames.DataFrame=ACROSS_INDI, WITHIN_POOL::DataFrames.DataFrame=WITHIN_POOL, WITHIN_INDI::DataFrames.DataFrame=WITHIN_INDI)
  ### test
  # par = [0.25, 0.25, 0.25, 0.25]
  # nLib = 10
  # nQTL = 5
  ### random sampling and summary statistics i.e CORRELATION, LOG10_RMSD, TRUE_POSITIVE_RATE, and FALSE_POSITIVE_RATE
  DATA_LEVELS = [ACROSS_POOL, ACROSS_INDI, WITHIN_POOL, WITHIN_INDI]
  CORRELATION = []
  LOG10_RMSD = []
  TRUE_POSITIVE_ID = []
  FALSE_POSITIVE_ID = []
  for i in 1:length(DATA_LEVELS)
    # i = 1
    LEVEL = DATA_LEVELS[i]
    n = convert(Int64, round(par[i]*nLib))
    if (n > 0)
      idx = rand(collect(1:size(LEVEL,1)), n)
      # println(idx)
      push!(CORRELATION, mean(dropmissing(LEVEL[idx,:]).CORRELATION))
      push!(LOG10_RMSD, mean(dropmissing(LEVEL[idx,:]).LOG10_RMSD))
      push!(TRUE_POSITIVE_ID, split(join(convert(Array{String}, dropmissing(LEVEL[idx,:]).TRUE_POSITIVE_ID), ";"), ";"))
      push!(FALSE_POSITIVE_ID, split(join(convert(Array{String}, dropmissing(LEVEL[idx,:]).FALSE_POSITIVE_ID), ";"), ";"))
    end
  end
  if (sum(isnan.(CORRELATION))>0) | (sum(ismissing.(CORRELATION))>0)
    SAMPLE_ABC(par=par, nLib=nLib, nQTL=nQTL)
  else
    OUT = [mean(CORRELATION[.!isnan.(CORRELATION)]),
           mean(LOG10_RMSD[.!isnan.(LOG10_RMSD)]),
           length(unique(TRUE_POSITIVE_ID))/nQTL,
           length(unique(FALSE_POSITIVE_ID))/(length(unique(TRUE_POSITIVE_ID))+length(unique(FALSE_POSITIVE_ID)))]
    return(OUT)
  end
end
### parallel computation using @distributed and prefixed with
### @sync to wait for the whole parallel computation to finish before runing the next line
OUT = SharedArrays.SharedArray{Float64,2}(size(PAR,1)*nREP, 4)
@time x = @sync @distributed for i = 1:size(OUT,1)
  idx_PAR = convert(Int64, i - (floor((i-1)/size(PAR,1))*size(PAR,1)))
  OUT[i,:] = SAMPLE_ABC(par=convert(Array{Float64,1},PAR[idx_PAR,:]), nLib=10)
end
### merge the parameters and the corresponding summary statistics from the simulated (sampled) data
MERGED = DataFrames.DataFrame(REP=repeat(collect(1:nREP), inner=size(PAR,1)),
                              ACROSS_POOL=repeat(PAR.ACROSS_POOL, outer=nREP),
                              ACROSS_INDI=repeat(PAR.ACROSS_INDI, outer=nREP),
                              WITHIN_POOL=repeat(PAR.WITHIN_POOL, outer=nREP),
                              WITHIN_INDI=repeat(PAR.WITHIN_INDI, outer=nREP),
                              CORRELATION=OUT[:,1],
                              LOG10_RMSD=OUT[:,2],
                              TRUE_POSITIVE_RATE=OUT[:,3],
                              FALSE_POSITIVE_RATE=OUT[:,4])
CSV.write(string("ABC_SAMPLING_OUTPUT_", nREP, "REPS.csv"), MERGED)
